### DISCUSSION TITLE
Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client

### DISCUSSION THREAD

**Discussion by Hakamate** [adonisjs/core#1305] [Help] [answered]
## [Send file with adonis server]

Hello everyone !

I'm just trying to do a youtube downloader feature

My Download method (Adonis) :

```
download({ request, response }) {
        let title = null
        const options = []
        const req = request.all()
        var link = req.params.URL

        const video = youtubedl(link, ["--format=bestaudio"], {cwd: __dirname});
       
        youtubedl.getInfo(link, options, function(err, info){
            if(err) throw err
            title = info.title
            const target = Helpers.publicPath(title + ".mp3")
    
            video.pipe(
                fs.createWriteStream(target)
            )
            video.on('end', function() {
                console.log(title, ' finished downloading !')
                response.attachment(target)
              })
        })
    }
```

My Axios call (Nuxt): 

```
await axios.post(`http://127.0.0.1:3333/api/download`, {params:{URL:url}})
        .catch(error => {
          console.log(error);
        })
        .then(response => console.log(response))
```

I'm trying to send a downloaded file to my nuxt client. But a problem occured. 
When i call my Adonis serveur, it responds directly, but doesn't wait the downloaded file, any Idea ? 

Do i have to set an header ? But how ?

**Comment by McSneaky** [comment, marked as answer]
Dealing with AdonisJS version 4?
Then this section will help: 
https://adonisjs.com/docs/4.0/response#_making_response
If with version 5, then there's example project that does quite the same:
https://gitlab.com/McSneaky/youtube-downloader/-/blob/master/start/routes.ts#L28

**Comment by Hakamate** [comment]
Thanks you @McSneaky 

I previously used `response.implicitEnd = false` but not a the beggining, that why it didn't working.
